File: src/cpp/fuerzaBruta.cpp
#include <string>
#include <iostream>
#include <fstream>
#include "include/json.hpp"
#include <tuple>
#include <cstdlib>
#include <vector>
#include <map>
#include <algorithm>

using namespace std;

// Para libreria de JSON.
using namespace nlohmann;
float mapValue(float value, float fromLow, float fromHigh, float toLow, float toHigh)
{
    return (value - fromLow) / (fromHigh - fromLow) * (toHigh - toLow) + toLow;
}
float calcular_error(tuple<int, int> a, tuple<int, int> b, json instance, int n, int m)
{
    float AX = mapValue(get<0>(a), 0, n - 1, instance["x"][0], instance["x"][int(instance["n"]) - 1]);
    float BX = mapValue(get<0>(b), 0, n - 1, instance["x"][0], instance["x"][int(instance["n"]) - 1]);
    float AY = mapValue(get<1>(a), 0, m - 1, instance["y"][0], instance["y"][int(instance["n"]) - 1]);
    float BY = mapValue(get<1>(b), 0, m - 1, instance["y"][0], instance["y"][int(instance["n"]) - 1]);

    float error = 0;
    for (int i = 0; i < instance["n"]; i++)
    {
        if (AX <= instance["x"][i] && instance["x"][i] <= BX)
        {
            float predicted_y = ((BY - AY) / (BX - AX)) * (float(instance["x"][i]) - AX) + AY;
            error += abs(float(instance["y"][i]) - predicted_y);
        }
    }
    return error;
}

map<vector<tuple<int, int>>, float> fuerza_bruta_recursiva(int n, int m, int N, json instance, int i, vector<tuple<int, int>> bp, float error_total, map<vector<tuple<int, int>>, float> &combinaciones)
{
    cout << "recursion" << endl;
    if (bp.size() == N && get<0>(bp[bp.size() - 1]) == m - 1)
    {

        combinaciones[bp] = error_total;
    }
    if (bp.size() == 0)
    {
        for (int z = 0; z < m; z++)
        {
            vector<tuple<int, int>> new_bp = {make_tuple(0, z)};
            fuerza_bruta_recursiva(m, n, N, instance, 0, new_bp, error_total, combinaciones);
        }
        for (int j = 0; j < m; j++)
        {
            for (int k = i + 1; k < n; k++)
            {
                vector<tuple<int, int>> new_bp = bp;
                new_bp.push_back(make_tuple(k, j));
                if (bp.size() != 0)
                {
                    float error = calcular_error(bp[bp.size() - 1], make_tuple(k, j), instance, n, m);
                    fuerza_bruta_recursiva(m, n, N, instance, k, new_bp, error_total + error, combinaciones);
                }
            }
        }
    }
    return combinaciones;
}

pair<vector<tuple<int, int>>, float> fuerza_bruta(int n, int m, int N, json instance)
{

    // map<vector<tuple<int,int>>,float> combinaciones = {};
    // fuerza_bruta_recursiva(n,m,N,instance,0, {}, 0.0 , combinaciones);
    // for (auto it = combinaciones.begin(); it != combinaciones.end(); ++it) {
    // #include <algorithm> // para std::min_element

    map<vector<tuple<int, int>>, float> combinaciones = {};
    fuerza_bruta_recursiva(n, m, N, instance, 0, {}, 0.0, combinaciones);

    auto min_it = combinaciones.begin();
    for (auto it = combinaciones.begin(); it != combinaciones.end(); ++it)
    {
        if (it->second < min_it->second)
        {
            min_it = it;
        }
    }

    return *min_it; // devuelve el par clave-valor con el valor mínimo
}
int main(int argc, char **argv)
{

    string instance_name = "data/titanium.json";
    cout << "Reading file " << instance_name << endl;
    ifstream input(instance_name);

    json instance;
    input >> instance;
    input.close();

    int K = instance["n"];
    int m = 6;
    int n = 6;
    int N = 5;

    // float prueba = calcular_error(make_tuple(0,0), make_tuple(2,0), instance,6,6);
    // cout << prueba << endl;
    // cout << K << endl;

    // Aca empieza la magia.

    // Ejemplo para guardar json.
    // Probamos guardando el mismo JSON de instance, pero en otro archivo.

    std::pair<std::vector<std::tuple<int, int>>, float> minimo = fuerza_bruta(n, m, N, instance);

    // Crea un objeto json
    nlohmann::json j;

    // Añade los datos al objeto json
    j["n"] = minimo.first.size();
    j["obj"] = minimo.second;

    // Divide el vector de tuplas en dos vectores x e y
    std::vector<float> x, y;
    for (const auto &tup : minimo.first)
    {
        x.push_back(std::get<0>(tup));
        y.push_back(std::get<1>(tup));
    }

    j["x"] = x;
    j["y"] = y;

    // Escribe el objeto json a un archivo
    std::ofstream o("minimo.json");
    o << j << std::endl;

    ofstream output("test_output.out");

    output << instance;
    output.close();

    return 0;
}

File: src/python/fuerzaBruta.py
import json
import numpy as np
import os
import time

from shared import calcular_error, plot_graph

def fuerza_bruta_recursiva(m, n, N, instance, i, bp, error_total, combinaciones, grid_x, grid_y):
    """
    Función recursiva para encontrar todas las combinaciones de breakpoints y calcular el error total para cada combinación.
    
    Parámetros:
        m (int): Número de puntos en la grilla a lo largo del eje x.
        n (int): Número de puntos en la grilla a lo largo del eje y.
        N (int): Número deseado de breakpoints.
        instance (dict): Instancia de datos en formato JSON.
        i (int): Índice de la fila actual en la grilla.
        bp (list): Lista de breakpoints actuales.
        error_total (float): Error acumulado hasta el momento.
        combinaciones (dict): Diccionario para almacenar las combinaciones de breakpoints y sus errores totales.
        grid_x (numpy.ndarray): Coordenadas x de la grilla.
        grid_y (numpy.ndarray): Coordenadas y de la grilla.
    
    Devuelve:
        tuple: Tupla que contiene la mejor combinación de breakpoints, su error total y el diccionario de combinaciones.
    """
    # CASO BASE:
        # Se han alcanzado N breakpoints incluyendo de la última columna:
    if len(bp) == N and bp[-1][0] == m-1:
        # Registra la combinación actual y su error total.
        combinaciones[tuple(bp)] = round(error_total, 3)
        return bp, error_total, combinaciones
    # Si es el primer breakpoint, llama recursivamente sobre todas las f(x), sin añadir error.
    if not bp:
        for z in range(m):
            new_bp = [(0,z)]
            fuerza_bruta_recursiva(m, n, N, instance, 0, new_bp, error_total, combinaciones, grid_x, grid_y)
    # Si ya hay breakpoints pero menos a los esperados (N):
    elif len(bp) < N: 
        # Itera sobre todas las filas y columnas para analizar c/combinación posible.
        for j in range(m):
            for k in range(i+1,n):
                # Crea una nueva lista de breakpoints añadiendo el punto actual (k, j).
                new_bp = bp + [(k, j)]
                if bp:
                    # Calcula el error con el nuevo punto.
                    error = calcular_error(bp[-1], (k, j), grid_x, grid_y, instance)
                    # Llama recursivamente para agregar el próximo breakpoint con el nuevo error total.
                    fuerza_bruta_recursiva(m, n, N, instance, k, new_bp, error_total + error, combinaciones, grid_x, grid_y)

    # Retorna la lista actual de breakpoints, el error total acumulado y el diccionario de combinaciones probadas.
    return bp, error_total, combinaciones

def fuerza_bruta(m, n, N, instance):
    # Crea la grilla.
    grid_x = np.linspace(min(instance["x"]), max(instance["x"]), num=m, endpoint=True)
    grid_y = np.linspace(min(instance["y"]), max(instance["y"]), num=n, endpoint=True)
    # Crea el diccionario con las soluciones factibles.
    combinaciones = {}
    fuerza_bruta_recursiva(m, n, N, instance, 0, [], 0, combinaciones, grid_x, grid_y)

    # Ordenamos el diccionario para devolver las mejores 5, según menor error.
    top_combinaciones = sorted(combinaciones.items(), key=lambda item: item[1])[:5]
    
    print(f"Top 5 Combinaciones de {len(combinaciones)}:")
    for idx, (comb, error) in enumerate(top_combinaciones, 1):
        print(f"{idx}: {comb} con error: {error}")
        # Extrae la mejor combinación.
    best_combination, min_error = top_combinaciones[0]

    # Conversión breakpoints-puntos de la grilla para best combination.
    best_x = [grid_x[x[0]] for x in best_combination]
    best_y = [grid_y[y[1]] for y in best_combination]

    # Construye el diccionario-solución
    solution = {
        'n': len(best_combination),
        'x': best_x,
        'y': best_y,
        'obj': min_error
    }
    return solution, min_error

def main():
    files = ['optimistic_instance.json'] #'aspen_simulation.json', 'ethanol_water_vle.json', 'titanium.json', 'optimistic_instance.json', 'toy_instance.json'
    reps = 1
    for filename in files:
        # Carga los datos desde la instancia JSON.
        instance_name = filename
        filename = "data/" + instance_name
        with open(filename) as f:
            instance = json.load(f)
        # Define los parámetros de la grilla y los breakpoints.
        for w in range(6,10):

            m = 10
            n = 10
            N = w
            
            for i in range(reps):
                start_time = time.time()

                # Obtener la solución utilizando programacion_dinamica
                solution, min_error = fuerza_bruta(m, n, N, instance)

                end_time = time.time()
                excecution_time = end_time - start_time
                total_excecution_time =+ excecution_time
            
            average_excecution_time = total_excecution_time/reps

            # Asegúrate de que el directorio exista
            solution_directory = 'data/solutions'
            if not os.path.exists(solution_directory):
                os.makedirs(solution_directory)

            # Exporta la solución a un archivo JSON
            solution_filename = os.path.join(solution_directory, f'solution_{instance_name}')
            try:
                with open(solution_filename, 'w') as f:
                    json.dump(solution, f)
                print(f'Solution exported to {solution_filename}')
            except Exception as e:
                print(f"Error al guardar la solución: {e}")

            # Gráfica la solución.
            print(w)
            plot_graph(instance_name, m, n, N, average_excecution_time, min_error, 'Fuerza Bruta', w)

if __name__ == "__main__":
    main()